client/X11/xf_cliprdr.c

```diff
@@ -1663,6 +1928,334 @@ static UINT xf_cliprdr_clipboard_file_range_failure(wClipboardDelegate* delegate
+				pos += direntry_len;
+			}
+			else if (index == 1)
+			{
+				struct stat stbuf;
+				stbuf.st_ino = node->parent_ino;
+				direntry_len = fuse_add_direntry(req, NULL, 0, "..", NULL, index);
+				if (pos + direntry_len > size)
+				{
+					break;
+				}
+				fuse_add_direntry(req, buf + pos, size - pos, "..", &stbuf, index);
+				pos += direntry_len;
+			}
+			else
+			{
+				child_ino = (size_t)ArrayList_GetItem(node->child_inos, index - 2);
+				ArrayList_Lock(clipboard->ino_list);
+				child_node =
+				    (xfCliprdrFuseInode*)ArrayList_GetItem(clipboard->ino_list, child_ino - 1);
+				ArrayList_Unlock(clipboard->ino_list);
+				// child_node must exists?
+				direntry_len = fuse_add_direntry(req, NULL, 0, child_node->name, NULL, index);
+				if (pos + direntry_len > size)
+				{
+					break;
+				}
+
+				struct stat stbuf;
+				stbuf.st_ino = child_node->ino;
+				fuse_add_direntry(req, buf + pos, size - pos, child_node->name, &stbuf, index);
+				pos += direntry_len;
+			}
+		}
+
+		fuse_reply_buf(req, buf, pos);
+		free(buf);
+	}
+	ArrayList_Unlock(node->child_inos);
+}
+
+static void xf_cliprdr_fuse_open(fuse_req_t req, fuse_ino_t ino, struct fuse_file_info* fi)
+{
+	xfCliprdrFuseInode* node;
+	xfClipboard* clipboard = (xfClipboard*)fuse_req_userdata(req);
+
+	ArrayList_Lock(clipboard->ino_list);
+	node = (xfCliprdrFuseInode*)ArrayList_GetItem(clipboard->ino_list, ino - 1);
+	ArrayList_Unlock(clipboard->ino_list);
+	if (!node)
+	{
+		fuse_reply_err(req, ENOENT);
+	}
+	else if ((node->st_mode & S_IFDIR) != 0)
+	{
+		fuse_reply_err(req, EISDIR);
+	}
+	else
+	{
+		/* Important for KDE to get file correctly*/
+		fi->direct_io = 1;
+		fuse_reply_open(req, fi);
+	}
+}
+
+static void xf_cliprdr_fuse_read(fuse_req_t req, fuse_ino_t ino, size_t size, off_t off,
+                                 struct fuse_file_info* fi)
+{
+	if (ino < 2)
+	{
+		fuse_reply_err(req, ENONET);
+		return;
+	}
+	xfCliprdrFuseInode* node;
+	xfClipboard* clipboard = (xfClipboard*)fuse_req_userdata(req);
+
+	ArrayList_Lock(clipboard->ino_list);
+	node = (xfCliprdrFuseInode*)ArrayList_GetItem(clipboard->ino_list, ino - 1);
+	ArrayList_Unlock(clipboard->ino_list);
+	if (!node)
+	{
+		fuse_reply_err(req, ENONET);
+		return;
+	}
+	else if ((node->st_mode & S_IFDIR) != 0)
+	{
+		fuse_reply_err(req, EISDIR);
+		return;
+	}
+
+	xfCliprdrFuseStream* stream = (xfCliprdrFuseStream*)calloc(1, sizeof(xfCliprdrFuseStream));
+
+	ArrayList_Lock(clipboard->stream_list);
+	stream->req = req;
+	stream->stream_id = clipboard->current_stream_id;
+	clipboard->current_stream_id++;
+	ArrayList_Add(clipboard->stream_list, stream);
+	ArrayList_Unlock(clipboard->stream_list);
+
+	UINT32 nPositionLow = (off >> 0) & 0xFFFFFFFF;
+	UINT32 nPositionHigh = (off >> 32) & 0xFFFFFFFF;
+
+	xf_cliprdr_send_client_file_contents(clipboard, stream->stream_id, node->lindex, nPositionLow,
+	                                     nPositionHigh, size);
```

Comment:
Can it happen that we have several overlapping read() requests in flight for a single file? 🤔